### PR TITLE
Fix doxygen warnings in utilities/ headers

### PR DESCRIPTION
**Pull request by karthikeyann**
Fixes parts of https://github.com/rapidsai/cudf/issues/9373
added missing documentation to fix doxygen warnings in multiple files cpp/include/cudf/utilities/
fixes 167 warnings

### CODE REVIEW

File: cpp/include/cudf/utilities/bit.hpp [hunk 1]
```diff
@@ -79,6 +93,7 @@ constexpr CUDF_HOST_DEVICE inline size_type intra_word_index(size_type bit_index
  *
  * @param bitmask The bitmask containing the bit to set
  * @param bit_index Index of the bit to set
+ * @return void
```
Comment (davidwendt) [inline cpp/include/cudf/utilities/bit.hpp:96]: Is this needed?

Comment (karthikeyann) [reply cpp/include/cudf/utilities/bit.hpp:96]: This was a bug in old doxygen version `1.8.18`. Version 1.8.20 (which we use) does not throw this error.

Comment (karthikeyann) [reply cpp/include/cudf/utilities/bit.hpp:96]:
```suggestion
```

File: cpp/include/cudf/utilities/bit.hpp [hunk 2]
```diff
@@ -94,6 +109,7 @@ CUDF_HOST_DEVICE inline void set_bit_unsafe(bitmask_type* bitmask, size_type bit
  *
  * @param bitmask The bitmask containing the bit to clear
  * @param bit_index The index of the bit to clear
+ * @return void
```
Comment (karthikeyann) [inline cpp/include/cudf/utilities/bit.hpp:112]:
```suggestion
```

File: cpp/include/cudf/utilities/span.hpp [hunk 3]
```diff
@@ -43,56 +45,129 @@ class span_base {
...
   constexpr span_base() noexcept {}
+  /**
+   * @brief Constructs a span from a pointer and a size.
+   *
+   * @param data Pointer to the first element in the span.
+   * @param size The number of elements in the span.
+   */
   constexpr span_base(pointer data, size_type size) : _data(data), _size(size) {}
   // constexpr span_base(pointer begin, pointer end) : _data(begin), _size(end - begin) {}
-  constexpr span_base(span_base const& other) noexcept = default;
```
Comment (davidwendt) [inline cpp/include/cudf/utilities/span.hpp:61]: Looks like this line was removed by mistake perhaps?

Comment (karthikeyann) [reply cpp/include/cudf/utilities/span.hpp:61]: The line is not removed. only the argument name `other` is removed.

File: cpp/include/cudf/utilities/span.hpp [hunk 4]
```diff
@@ -240,47 +330,129 @@ namespace detail {
...
+   * @return The flattened index of the element at the specified 2D position
+   */
   static constexpr size_t flatten_index(size_t row, size_t column, size_type size) noexcept
   {
     return row * size.second + column;
   }
 
+  /**
+   * @brief Returns a reference to the row-th element of the sequence.
+   * The behavior is undefined if row is out of range (i.e., if it is greater than or equal to
```
Comment (davidwendt) [inline cpp/include/cudf/utilities/span.hpp:396]:
```suggestion
   * @brief Returns a reference to the row-th element of the sequence.
   *
   * The behavior is undefined if row is out of range (i.e., if it is greater than or equal to
```
```suggestion
   * @brief Returns a reference to the row-th element of the sequence.
   * The behavior is undefined if row is out of range (i.e., if it is greater than or equal to
```

File: cpp/include/cudf/utilities/span.hpp [hunk 5]
```diff
@@ -240,47 +330,129 @@ namespace detail {
...
+   * @return A reference to the row-th element of the sequence, i.e., `data()[row]`
+   */
   constexpr RowType<T, dynamic_extent> operator[](size_t row) const
   {
     return {this->data() + flatten_index(row, 0, this->size()), this->size().second};
   }
 
+  /**
+   * @brief Returns a reference to the first element in the span.
+   * Calling front on an empty span results in undefined behavior.
```
Comment (davidwendt) [inline cpp/include/cudf/utilities/span.hpp:409]:
```suggestion
   * @brief Returns a reference to the first element in the span.
   *
   * Calling front() on an empty span results in undefined behavior.
```
```suggestion
   * @brief Returns a reference to the first element in the span.
   * Calling front on an empty span results in undefined behavior.
```

Comment (davidwendt) [inline cpp/include/cudf/utilities/span.hpp:409]:
```suggestion
   * @brief Returns a reference to the first element in the span.
   *
   * Calling front() on an empty span results in undefined behavior.
```
```suggestion
   * @brief Returns a reference to the first element in the span.
   * Calling front on an empty span results in undefined behavior.
```

File: cpp/include/cudf/utilities/span.hpp [hunk 6]
```diff
@@ -240,47 +330,129 @@ namespace detail {
...
+   *
+   * @return Reference to the first element in the span
+   */
   [[nodiscard]] constexpr RowType<T, dynamic_extent> front() const { return (*this)[0]; }
+  /**
+   * @brief Returns a reference to the last element in the span.
+   * Calling last on an empty span results in undefined behavior.
```
Comment (davidwendt) [inline cpp/include/cudf/utilities/span.hpp:416]:
```suggestion
   * @brief Returns a reference to the last element in the span.
   *
   * Calling back() on an empty span results in undefined behavior.
```
```suggestion
   * @brief Returns a reference to the last element in the span.
   * Calling last on an empty span results in undefined behavior.
```

Comment (davidwendt) [inline cpp/include/cudf/utilities/span.hpp:416]:
```suggestion
   * @brief Returns a reference to the last element in the span.
   *
   * Calling back() on an empty span results in undefined behavior.
```
```suggestion
   * @brief Returns a reference to the last element in the span.
   * Calling last on an empty span results in undefined behavior.
```

**Review by davidwendt — CHANGES_REQUESTED** [review]

**Review by davidwendt — COMMENTED** [review]
Looks good.

**Review by rgsl888prabhu — APPROVED** [review]

File: cpp/include/cudf/utilities/span.hpp [hunk 7]
```diff
@@ -31,6 +31,8 @@
 
 namespace cudf {
 
+/// A constant of type std::size_t that is used to differentiate std::span of static and dynamic
+/// extent
```
Comment (davidwendt) [inline cpp/include/cudf/utilities/span.hpp:35]:
```suggestion
/// A constant used to differentiate std::span of static and dynamic extent
```
I don't think we need to mention the type in the comment.

File: cpp/include/cudf/utilities/span.hpp [hunk 8]
```diff
@@ -43,56 +45,129 @@ class span_base {
...
+   * @brief Returns a reference to the first element in the span.
+   * Calling front on an empty span results in undefined behavior.
+   *
+   * @return Reference to the first element in the span
+   */
   constexpr reference front() const { return _data[0]; }
   // not noexcept due to undefined behavior when size = 0
+  /**
+   * @brief Returns a reference to the last element in the span.
+   * Calling last on an empty span results in undefined behavior.
```
Comment (davidwendt) [inline cpp/include/cudf/utilities/span.hpp:89]:
```suggestion
   * @brief Returns a reference to the last element in the span.
   *
   * Calling back() on an empty span results in undefined behavior.
```
Missing blank line after `@brief` appears multiple times in this PR.

**Review by davidwendt — APPROVED** [review]

**Comment by karthikeyann** [comment]
@gpucibot merge

**Pull request merged**